### ISSUE TITLE
Example for enum from P4-16 documentation results in error

### ISSUE THREAD

**Issue by c-bg** [p4lang/p4c#1534]
Putting
```
#include <core.p4>
#include <v1model.p4>

// ...

enum bit<8> IP_Prot_Num
{
    ICMP = 0x01,
    TCP = 0x06
    // ...
}

// ...
```
in my p4-file results in 
```
syntax error, unexpected BIT
enum bit
     ^^^
```
even though that is analogous to the example in the documentation:
```
enum bit<16> EtherType {
    VLAN = 0x8100,
    QINQ = 0x9100,
    MPLS = 0x8847,
    IPV4 = 0x0800,
    IPV6 = 0x86dd
}
```
(Section: 7.2.1. Enumeration types, Page: 41; from https://github.com/p4lang/p4-spec/tree/master/p4-16/spec)

Does someone have an idea what is happening here?

**Comment by antoninbas** [comment]
Are you using a recent compiler version?

**Comment by mbudiu-vmw** [comment]
Do you have the following file in your testsuite: `testdata/p4_16_samples/serenum.p4`?
If you don't it means that you are using an old version of the compiler.
If you do and that test passes there is something else wrong with your P4 program.

**Comment by c-bg** [comment]
The p4c version is 0.5, since I tried out the program in the tutorial vm (https://github.com/p4lang/tutorials/tree/master/vm). If you do not get the same error, I assume this issue was indeed resolved in the newer versions, and am very sorry for the inconvenience!

**Comment by mbudiu-vmw** [comment]
We don't do a good job of incrementing the version so it's not a very good indicator.
Your safest bet would be to build the compiler from the sources in this repository.

**Comment by jafingerhut** [comment]
But when doing so, be aware that if you are running code examples in the p4lang/tutorials repository, those are often tested with particular versions of the open source projects, and those examples often stop working with later versions of the tools.  You may want to use a different VM, or at least create a new directory on the same machine, for the latest version of the p4c source code to try it out there, if you want to keep using the tutorials.

**Comment by c-bg** [comment]
Thank you all so much for the fast and helpful replies! I appreciate it a lot.
Regarding building from source: I tried doing that before writing the previous reply, and stumbled upon an error that seems as if I should be capable of solving it myself...
When running
```
cmake .. [-DCMAKE_BUILD_TYPE=RELEASE|DEBUG] [-DCMAKE_INSTALL_PREFIX=<path>] [-DENABLE_DOCS=ON (default off)] [-DENABLE_P4RUNTIME_TO_PD=OFF (default on)] [-DENABLE_PROTOBUF_STATIC=OFF (default on)] [-DENABLE_GC=OFF (default on)]

```
as specified in https://github.com/p4lang/p4c, I get all sorts of CMake Errors, depending on which characters I escape.
Without escaping anything:

```
bash: syntax error near unexpected token `('
```
when escaping `(` and `)`:

```
CMake Error: The source directory "/home/p4/p4c/build/[-DCMAKE_BUILD_TYPE=RELEASE" does not exist.

```
Escaping even more characters did not help.
This is where I am currently stuck, so if anyone happens to know the solution by chance, feel free to share it, otherwise I am going to continue researching possible error sources.

btw, thanks for the tip regarding the vm. To be on the safe side, I had indeed already taken a snapshot - anyway it is still good to know that the tutorials might not work with newer versions.

**Comment by mbudiu-vmw** [comment]
`[ ]` indicate optional arguments. Try just `cmake ..`

**Comment by c-bg** [comment]
ah, yes, of course, that makes a lot of sense. `cmake ..` succeeded partially, it concluded with errors. `CMakeError.log` contains:

```
Determining if the pthread_create exist failed with the following output:
Change Dir: /home/p4/p4c/build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/make" "cmTC_70157/fast"
/usr/bin/make -f CMakeFiles/cmTC_70157.dir/build.make CMakeFiles/cmTC_70157.dir/build
make[1]: Entering directory '/home/p4/p4c/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_70157.dir/CheckSymbolExists.c.o
/usr/bin/cc     -o CMakeFiles/cmTC_70157.dir/CheckSymbolExists.c.o   -c /home/p4/p4c/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c
Linking C executable cmTC_70157
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_70157.dir/link.txt --verbose=1
/usr/bin/cc       CMakeFiles/cmTC_70157.dir/CheckSymbolExists.c.o  -o cmTC_70157 -rdynamic
CMakeFiles/cmTC_70157.dir/CheckSymbolExists.c.o: In function `main':
CheckSymbolExists.c:(.text+0x16): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
CMakeFiles/cmTC_70157.dir/build.make:97: recipe for target 'cmTC_70157' failed
make[1]: *** [cmTC_70157] Error 1
make[1]: Leaving directory '/home/p4/p4c/build/CMakeFiles/CMakeTmp'
Makefile:126: recipe for target 'cmTC_70157/fast' failed
make: *** [cmTC_70157/fast] Error 2

File /home/p4/p4c/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:
/* */
#include <pthread.h>

int main(int argc, char** argv)
{
  (void)argv;
#ifndef pthread_create
  return ((int*)(&pthread_create))[argc];
#else
  (void)argc;
  return 0;
#endif
}

Determining if the function pthread_create exists in the pthreads failed with the following output:
Change Dir: /home/p4/p4c/build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/make" "cmTC_9660e/fast"
/usr/bin/make -f CMakeFiles/cmTC_9660e.dir/build.make CMakeFiles/cmTC_9660e.dir/build
make[1]: Entering directory '/home/p4/p4c/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_9660e.dir/CheckFunctionExists.c.o
/usr/bin/cc    -DCHECK_FUNCTION_EXISTS=pthread_create   -o CMakeFiles/cmTC_9660e.dir/CheckFunctionExists.c.o   -c /usr/share/cmake-3.5/Modules/CheckFunctionExists.c
Linking C executable cmTC_9660e
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_9660e.dir/link.txt --verbose=1
/usr/bin/cc   -DCHECK_FUNCTION_EXISTS=pthread_create    CMakeFiles/cmTC_9660e.dir/CheckFunctionExists.c.o  -o cmTC_9660e -rdynamic -lpthreads
/usr/bin/ld: cannot find -lpthreads
collect2: error: ld returned 1 exit status
CMakeFiles/cmTC_9660e.dir/build.make:97: recipe for target 'cmTC_9660e' failed
make[1]: *** [cmTC_9660e] Error 1
make[1]: Leaving directory '/home/p4/p4c/build/CMakeFiles/CMakeTmp'
Makefile:126: recipe for target 'cmTC_9660e/fast' failed
make: *** [cmTC_9660e/fast] Error 2


```
Accordingly,

```
make -j4
make -j4 check

```
both failed with
```
make: *** No targets specified and no makefile found.  Stop.

```
and
```
make: *** No rule to make target 'check'.  Stop.

```
respectively.

I guess this yet again is an error which has a fairly obvious reason that I somehow cannot see right now. Sorry for the inconveniences.
Anyway, out of curiosity, I reran my previously mentioned p4 program, but still got the same error, so I guess that the above error indeed prevented the p4c command from getting updated.

**Comment by mbudiu-vmw** [comment]
Have you installed all the required dependencies as described in the README?

**Comment by c-bg** [comment]
yes, including the protobuf in version 3.2.0 as recommended for Ubuntu. 
In the meantime, I've also tried the solution proposed in #1495, namely
```
git clone https://github.com/jafingerhut/p4-guide
./p4-guide/bin/install-p4dev.sh
```
the only difference being that I tried this solution out in the VM instead of on a totally new install (since earlier versions worked as well).
However, the issue persists.
I might try this approach on a fresh Ubuntu VM.

**Comment by antoninbas** [comment]
I'm pretty sure that this is not an error and that cmake actually succeeds. You don't need to worry about the contents of CMakeError.log in this case. What happens is that cmake checks for usability of pthreads, which fail, then it falls back to pthread. You can just move on and do make in the build directory.
If you are convinced that cmake actually failed, please share the stdout contents with us first, not CMakeError.log.

**Comment by c-bg** [comment]
Thank you all for your help and patience! 

I've now got a different VM with a more recent version of p4c (it includes the file `/p4c/testdata/p4_16_samples/serenum.p4`, something that was mentioned earlier in this discussion), and everything works now.
I am sorry to have submitted a non-issue, and to have asked some quite ridiculous questions (the answer to the `cmake ..` questions is in hindsight so obvious that it hurts...). I've learnt a lot.
So, in conclusion, if anyone decides to use the tutorials VM - the tutorials can be solved in it, so it's a good start, but if you want to develop your own code, it is (*obviously*) crucial to get the latest version of everything. -ahem-

Thanks a lot again and have a nice day.
